File: services/customers-service/src/test/java/br/com/deltaglobalbank/customers/domain/customer/valueobjects/CpfTests.java
package br.com.deltaglobalbank.customers.domain.customer.valueobjects;

import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;
import org.junit.jupiter.params.provider.ValueSource;

class CpfTests {

    @ParameterizedTest
    @ValueSource(strings = {"11144477735", "111.444.777-35"})
    void mustAcceptAValidCpf(String validCpf) {
        assertDoesNotThrow(() -> new Cpf(validCpf));
    }

    @ParameterizedTest
    @ValueSource(strings = {
        "11144477700",
        "11144477734",
        "11111111111",
        "123",
        "",
    })
    void mustRejectAnInvalidCpf(String invalidCpf) {
        assertThrows(IllegalArgumentException.class, () -> new Cpf(invalidCpf));
    }

    @ParameterizedTest
    @CsvSource({"111.444.777-35, 11144477735", "111444777-35, 11144477735"})
    void normalizesCpf(String input, String expected) {
        assertEquals(expected, new Cpf(input).value());
    }

    @Test
    void normalizesAMaskedCpfToDigitsOnly() {
        Cpf cpf = new Cpf("111.444.777-35");
        assertEquals("11144477735", cpf.value());
    }

    @Test
    void rejectsCpfWithInvalidCheckDigit() {
        assertThrows(IllegalArgumentException.class, () -> new Cpf("11144477700"));
    }

    @Test
    void rejectsCpfWithValidFirstButInvalidSecondCheckDigit() {
        assertThrows(IllegalArgumentException.class, () -> new Cpf("11144477734"));
    }

    @Test
    void rejectsCpfWithAllRepeatedDigits() {
        assertThrows(IllegalArgumentException.class, () -> new Cpf("11111111111"));
    }

    @Test
    void rejectsCpfWithWrongLength() {
        assertThrows(IllegalArgumentException.class, () -> new Cpf("123"));
    }
}

File: services/customers-service/src/test/kotlin/br/com/deltaglobalbank/customers/domain/customer/valueobjects/CpfTests.kt
package br.com.deltaglobalbank.customers.domain.customer.valueobjects

import org.junit.jupiter.api.Assertions.assertDoesNotThrow
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource
import org.junit.jupiter.params.provider.ValueSource
import org.junit.jupiter.api.Test
import kotlin.test.assertFailsWith

class CpfTests {

    @ParameterizedTest
    @ValueSource(strings = ["11144477735", "111.444.777-35"])
    fun `must accept a valid cpf`(validCpf: String) {
        assertDoesNotThrow { Cpf(validCpf) }
    }

    @ParameterizedTest
    @ValueSource(strings = [
        "11144477700",
        "11144477734",
        "11111111111",
        "123",
        "",
    ])
    fun `must reject an invalid cpf`(invalidCpf: String) {
        assertThrows(IllegalArgumentException::class.java) { Cpf(invalidCpf) }
    }

    @ParameterizedTest
    @CsvSource("111.444.777-35, 11144477735", "111444777-35, 11144477735")
    fun `normalizes cpf`(input: String, expected: String) {
        assertEquals(expected, Cpf(input).value)
    }

    @Test
    fun `normalizes a masked cpf to digits only`() {
        val cpf = Cpf("111.444.777-35")
        assertEquals("11144477735", cpf.value)
    }

    @Test
    fun `rejects cpf with invalid check digit`() {
        assertFailsWith<IllegalArgumentException> { Cpf("11144477700") }
    }

    @Test
    fun `rejects cpf with valid first but invalid second check digit`() {
        assertFailsWith<IllegalArgumentException> { Cpf("11144477734") }
    }

    @Test
    fun `rejects cpf with all repeated digits`() {
        assertFailsWith<IllegalArgumentException> { Cpf("11111111111") }
    }

    @Test
    fun `rejects cpf with wrong length`() {
        assertFailsWith<IllegalArgumentException> { Cpf("123") }
    }
}